core/bcast/bcast.go

```diff
@@ -110,6 +110,14 @@ func (b Broadcaster) Broadcast(ctx context.Context, duty core.Duty, set core.Sig
 		return err
 
 	case core.DutyBuilderRegistration:
+		slot, err := firstSlotInCurrentEpoch(ctx, b.eth2Cl)
+		if err != nil {
+			return errors.Wrap(err, "calculate first slot in epoch")
+		}
+
+		// Use first slot in current epoch for accurate delay calculations while submitting builder registrations.
+		// This is because builder registrations are submitted in first slot of every epoch.
+		duty.Slot = slot
```

Comment:
I'm wondering about an alternate approach:

We can probably do the slot from timestamp conversion at the validatorapi component: https://github.com/ObolNetwork/charon/blob/main/core/validatorapi/validatorapi.go#L522

This will create the accurate slot and initialize the duty with the correct slot (NewBuilderRegistrationDuty) starting at validatorapi itself

Reply:
yeah thought about this too, we need to do this at two places, first is here in validatorapi when VC submits registration and second is where we submit pre-gen registrations. Currently we use slot 0 for pregens and slot 1 for VC submissions which helps us in differentiating both types of registrations. 
Wanted to keep the differentiating part which is why I did this just before we submit in bcast and where we log/report the delays.